rand_seeder/src/sip.rs

```diff
@@ -169,6 +237,23 @@ impl<S: Sip> Hasher<S> {
         self.ntail = length - needed;
         self.tail = unsafe { u8to64_le(msg, needed, self.ntail) };
     }
+    
+    /// Finish writes and convert the core into an RNG
+    pub fn make_rng(self) -> SipRng {
+        let mut state = self.state;
+
+        // Finish
+        let b: u64 = ((self.length as u64 & 0xff) << 56) | self.tail;
+
+        state.v3 ^= b;
+        S::c_rounds(&mut state);
+        state.v0 ^= b;
+        
+        // This is supposed to be d - c rounds (here 4 - 2 = 2)
+        S::c_rounds(&mut state);
+        
+        SipRng::from_state(state)
```

Comment:
Where does this come from and how does it relate to the original algorithm?

Reply:
This is the original: https://github.com/rust-lang/rust/blob/master/src/libcore/hash/sip.rs#L334

It looks like I missed the tweak to `state.v2` before the end rounds — though the state is tweaked before the *next* two rounds anyway. Might be better with an extra tweak then but not too important probably.

I originally tried to make the first output match the initial one (or the first two from the 128-bit output version), but this wasn't so easy to turn into a nice RNG (because one version mutates two different parts of the state, and the other doesn't mutate any state between these four rounds).